lib/badge.rb

```diff
@@ -0,0 +1,100 @@
+class Badge
+  include DSL
+
+  attr_reader :title, :description, :goal, :rank, :user
+
+  def initialize(user)
+    @user = user
+    describe_context
+    @bestowment = check_bestowment(@rank)
+  end
+
+  def check_bestowment(rank = nil)
+    bestowment = Bestowment.where(
+      badge_id: self.class,
+      user: @user
+    )
+    bestowment = bestowment.where(rank: rank) if has_progress?
+    bestowment.first
+  end
+
+  def bestowed(rank = nil)
+    if has_progress?
+      state = self.class.const_get("Rank#{rank}")
+    else
+      state = self
+    end
+    Bestowment.create(
+      badge_id: self.class,
+      rank: rank,
+      user: @user,
+      bestowed_at: DateTime.now,
+      title: state.title,
+      description: state.description
+    )
+  end
+
+  def progress
+    if self.class.progress.nil?
+      nil
+    else
+      @progress ||= instance_eval(&self.class.progress)
+    end
+  end
+
+  def earned?
+    return false unless @goal
+    if progress.nil?
+      instance_eval(&@goal)
+    else
+      progress > @goal
+    end
+  end
+
+  def run
+    # bestowed badge if all ranks was earned
+    bestowed(@rank) if earned? && @bestowment.nil?
+    # bestowed previous rank if it not earned
+    prev = @rank - 1
+    if prev.positive? && has_progress? && check_bestowment(prev).nil?
+      bestowed(@rank - 1)
+    end
+  end
+
+  def has_progress?
+    self.class.ranks.present?
+  end
+
+  private
+
+  def describe_context
+    @rank = 0
+    stop = false
+    ranks = self.class.ranks
+    if ranks.blank?
+      @title = self.class.title
+      @description = self.class.description
+      @goal = self.class.goal
+    else
+      ranks.each do |rank|
+        break if stop
+        state = self.class.const_get(rank)
+        current_goal = get_goal_result(state.goal)
+        next unless progress < current_goal
+        @rank = state.rank
+        @title = state.title
+        @description = state.description
+        @goal = current_goal
+        stop = true
+      end
+    end
+  end
```

Comment:
This logic seems a bit heavy, and generally having both a `break` and a `next` in a loop is a sign you've gone too far.  Perhaps easier would be to provide a `Badge#previous_rank` method returning the badge class of the previous rank, so all you need to do is:

1. Check `Bestowment.find_by(badge_id: self.previous_rank.class.name, user: user).bestowed?`
2. If true: `Bestowment.where(badge_id: self.class.name, user: user).first_or_initialize`, generate the `progress` and `save!`
3. If false: return